exp/ingest/verify/main.go

```
@@ -0,0 +1,220 @@
+//      stored in your app. For example, if you only store accounts, all other
+//      ledger entry types should be ignored (return ignore = true).
+//   1. In a loop, get entries from history archive by calling GetEntries()
+//      and Write() your version of entries found in the batch (in any order).
+//   2. When GetEntries() return no more entries, call Verify with a number of
+//      entries in your storage (to find if some extra entires exist in your
+//      storage).
+// Functions will return StateError type if state is found to be incorrect.
+// Check Horizon for an example how to use this tool.
+type StateVerifier struct {
+	StateReader *io.SingleLedgerStateReader
+	// TransformFunction transforms (or ignores) ledger entries streamed from
+	// checkpoint buckets to match the form added by `Write`. Read
+	// TransformLedgerEntryFunction godoc for more information.
+	TransformFunction TransformLedgerEntryFunction
+
+	readEntries  int
+	wroteEntries int
+	readingDone  bool
+
+	currentEntries map[string]xdr.LedgerEntry
+}
+
+// GetLedgerKeys returns `count` ledger keys from history buckets storing
+// actual entries in cache to compare in Write.
+func (v *StateVerifier) GetLedgerKeys(count int) ([]xdr.LedgerKey, error) {
+	keys := make([]xdr.LedgerKey, 0, count)
+	if len(v.currentEntries) > 0 {
+		var entry xdr.LedgerEntry
+		for _, e := range v.currentEntries {
+			entry = e
+			break
+		}
+
+		// Ignore error as StateError below is more important
+		entryString, _ := xdr.MarshalBase64(entry)
+		err := StateError(errors.Errorf(
+			"Entries (%d) not found locally, example: %s",
+			len(v.currentEntries),
+			entryString,
+		))
+		return keys, err
+	}
+	v.currentEntries = make(map[string]xdr.LedgerEntry)
+
+	for count > 0 {
+		entryChange, err := v.StateReader.Read()
+		if err != nil {
+			if err == stdio.EOF {
+				v.readingDone = true
+				return keys, nil
+			}
+			return keys, err
+		}
+
+		entry := entryChange.MustState()
```

Is the reason we can use this `Must` function that will panic if `entryChange` doesn't contain a `State` type, because the `StateReader` will only ever return `State`s and not other things like `Create`, `Removed`, etc? Would if be safer to do something similar to the `ignore` code on line 89 where if a `entryChange` was another type of change we'd ignore it?

---

The only reason is that `StateReader.Read()` should always return `State` type. It saves us a few lines of code (error handling) but I think that in such case we should `panic` anyway because this is unexpected behaviour. If `StateReader` returns different type than `State` it means that there's a huge bug in `StateReader` or go compiler (yeah, we've seen them in the past: https://github.com/golang/go/issues/32560) and the application should basically stop working instead of keep processing invalid data.